src/contracts/deposit/leveldb/leveldb.service.ts

```diff
@@ -0,0 +1,185 @@
+   *
+   * @returns {Promise<void>} A promise that resolves when the database is successfully initialized.
+   * @private
+   */
+  private async setupLevel() {
+    this.db = new Level(await this.getDBDirPath(), {
+      valueEncoding: 'json',
+    });
+    await this.db.open();
+  }
+
+  /**
+   * Fetches and constructs the cache directory path for the current blockchain network.
+   *
+   * @returns {Promise<string>} A promise that resolves to the full path of the network-specific cache directory.
+   * @private
+   */
+  private async getDBDirPath(): Promise<string> {
+    const chainId = await this.providerService.getChainId();
+    const networkDir = `chain-${chainId}`;
+
+    return join(this.cacheDir, networkDir);
+  }
+
+  /**
+   * Asynchronously retrieves deposit events and headers from the database.
+   * Iterates through entries starting with 'deposit:' to collect data and fetches headers stored under 'header'.
+   * Handles errors by logging and returning default cache values.
+   *
+   * @returns {Promise<{data: VerifiedDepositEvent[], headers: VerifiedDepositEventsCacheHeaders}>} Cache data and headers.
+   * @public
+   */
+  public async getEventsCache(): Promise<{
+    data: VerifiedDepositEvent[];
+    headers: VerifiedDepositEventsCacheHeaders;
+  }> {
+    try {
+      const stream = this.db.iterator({ gte: 'deposit:', lte: 'deposit:\xFF' });
+
+      const data: VerifiedDepositEvent[] = [];
+
+      for await (const [, value] of stream) {
+        data.push(this.parseDepositEvent(value));
+      }
+      const headers: VerifiedDepositEventsCacheHeaders = JSON.parse(
+        await this.db.get('headers'),
+      );
+
+      return { data, headers };
+    } catch (error: any) {
+      if (error.code === 'LEVEL_NOT_FOUND') return this.cacheDefaultValue;
+      throw error;
+    }
+  }
+
+  /**
+   * Generates a deposit key string based on a given number.
+   * The number is checked to ensure it falls within a valid range (from 0 up to MAX_DEPOSIT_COUNT).
+   * If the number is out of bounds, an error is thrown.
+   * The method creates a buffer, writes the number to the buffer in big-endian format,
+   * and returns a deposit key string that includes the hexadecimal representation of the number.
+   *
+   * @param {number} number - The number used to generate the deposit key.
+   * @returns {string} The deposit key in the format 'deposit:XXXX', where 'XXXX' is the hexadecimal representation of the number.
+   * @throws {Error} If the number is less than 0 or greater than MAX_DEPOSIT_COUNT.
+   * @private
+   */
+  private generateDepositKey(number: number): string {
+    if (number < 0 || number > MAX_DEPOSIT_COUNT) {
+      throw new Error(
+        `Deposit count is out of the valid range (0 to ${MAX_DEPOSIT_COUNT}) received ${number}`,
+      );
+    }
+    const index = Buffer.alloc(4);
+    index.writeUInt32BE(number, 0);
+    return `deposit:${index.toString('hex')}`;
```

Comment:
Do we need the 'deposit:' prefix? Is it added in case there are different values in the same cache, not just deposit values?

Let's rename generateDepositKey to generateCacheKey. DepositKey sounds like a public key.

Reply:
got it.
generateDepositCacheKey is fine